### ISSUE TITLE
Strange instantiation limit behavior with where-clauses

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#20575]
The following is a sort of strange corner-case bug that likely doesn't need any attention in the near term, but I thought it would be good to document while I have it on my mind.

Consider the following program which uses both where-clauses and the `fn exempt instantiation limit` pragma:
```chapel
// param to select between two implementations of foo at compile time
config param ZeroIndexed = false;

// an implementation of foo that is not exempt from instantiation limits
proc foo(param n) where ZeroIndexed == true {
    var a : [0..<n] int;
    // ...
}

// a different implementation of foo that is exempt
pragma "fn exempt instantiation limit"
proc foo(param n) where ZeroIndexed == false {
    var a : [1..n] int;
    // ...
}

// instantiate 'foo' one too many times (the limit is 512)
for param n in 1..513 do foo(n);
```

When compiling the above program with `-s ZeroIndexed=false`, the following error message is emitted:
```
inst_limit_bug.chpl:5: error: Function 'foo' has been instantiated too many times
note:   If this is intentional, try increasing the instantiation limit from 512
```

And when compiling with `-s ZeroIndexed=true`, the program compiles and runs without any errors.

In other words, its as if the `pragma "fn exempt instantiation limit"` were applied to the zero-indexed version of foo rather than the non-zero-indexed version. Note that the line number on the error message is also incorrect.

**Comment by bradcray** [comment]
This is truly bizarre.  I have verified that you aren't hallucinating (or that we both are).

**Comment by jeremiah-corrado** [comment]
Yeah, it's quite strange. It came up while I was working on this PR: https://github.com/chapel-lang/chapel/pull/20508, because some of the internal IO functions (ex: `_write_one_internal`) are exempt from instantiation limits, and the where-clauses that I added to some of channel's methods were preventing the exemptions from going up the call-chain. 

So for example, top-level `writeln` didn't have an instantiation limit (because it eventually calls `_write_one_internal`). Then I added some where-clauses to select between implementations of `channel.write`, and the instantiation limit somehow no longer made it up to top-level `writeln`.

But I think the code I posted above summarizes the issue in its most fundamental form.

**Comment by bradcray** [comment]
Tagging @stonea in case he hasn't seen this and would be curious to look into it as the pragma's author (though it also wouldn't surprise me if the cause were something deeper and longer-standing).

**Comment by stonea** [comment]
It might be worth digging under this rock to see if there's some nasty underlying bug there.

For one reason or another we must be passing in the `where ZeroIndexed == true` overload of `foo` into [`trackInstantiationsForFn`](https://cs.github.com/chapel-lang/chapel/blob/ec6ab8c626b44bbfeb08fcd7b65d0c909ebd2b4a/compiler/resolution/generics.cpp#L181) which keeps count of how often a function has been instantiated:

For whatever it's worth I'll wildly speculate on why that might be:

* Perhaps we call that function before we've resolved what overload of the function applies so we just pick one arbitrarily to serve as the representative for all of these that we're going to count. 
* Perhaps this function is getting called for "failed" instantiations too (in that maybe we first try to "instantiate" the `where ZeroIndexe == true` overload of the function, track it with the `trackIntantiationsForFn` function, and then later realize that this isn't the right instantiation and move on).

In it's the second of these then maybe there's an underlying issue of us hitting the instantiation limit sooner than we should (like we'd hit it twice as fast for a function with two `where` based overloads)?  But at this point it's pure speculation.

I have to say: in terms of functionality for `pragma "fn exempt instantiation limit"` I'm not overly concerned about this. I wasn't thinking that was meant to be a user-facing feature (although maybe it should be?) and in practice if we use it in our standard library I'd expect we'd apply it to all overloads of the relevant function anyway.

Still. This is an odd one.

**Comment by stonea** [comment]
> For whatever it's worth I'll wildly speculate on why that might be:
> 
> * Perhaps we call that function before we've resolved what overload of the function applies so we just pick one arbitrarily to serve as the representative for all of these that we're going to count.
> * Perhaps this function is getting called for "failed" instantiations too (in that maybe we first try to "instantiate" the `where ZeroIndexe == true` overload of the function, track it with the `trackIntantiationsForFn` function, and then later realize that this isn't the right instantiation and move on).

Playing around with the example a little bit if it is one of these I think it's more likely the first than the second.  If I move the `pragma "fn exempt instantiation limit"` from the `ZeroIndexed == false` case to the `ZeroIndexed == true` then I won't get the "instantiated too many times" error (unexpected) and if I remove it all together I do (as would be expected).

IOW the pragma will get applied even though that's the not the overload of the function that, ultimately, we resolve.

**Comment by bradcray** [comment]
> I wasn't thinking that was meant to be a user-facing feature

I agree with this.
